### PR TITLE
194 region stencils

### PR DESCRIPTION
**Pull request by andrewcoughtrie**
Implementation of REGION stencil psylayer generation.

### CODE REVIEW

**Comment by andrewcoughtrie** [comment]
Implemented the REGION stencil tests and turned on the functionality. I've tested it against the algorithm/kernels that use the stencil in LFRic, it generates the correct code for the stencil. It isn't currently possible to test the code fully to running as there is additional functionality required for the kernels in question. 

When merging master into the branch some changes to the files in the external directory have appeared, I'm not sure if this is an issue or not.

The codecov/project check doesn't appear to pass but on closer inspection it looks like codecov is just being silly again.

File: doc/user_guide/dynamo0p3.rst [hunk 1]
```diff
@@ -1101,9 +1102,9 @@ Stencil metadata is written in the following format::
 
   STENCIL(type)
 
-where ``type`` may be one of ``X1D``, ``Y1D``, ``XORY1D``,
-``CROSS`` or ``CROSS2D``.  As the stencil ``extent`` (the maximum distance from the
-central cell that the stencil extends) is not provided in the metadata,
```
Comment (arporter) [inline doc/user_guide/dynamo0p3.rst:1106]: Please could you add an example of a region stencil, as done for the X1D and Cross stencils below.

File: src/psyclone/domain/lfric/lfric_arg_descriptor.py [hunk 2]
```diff
@@ -108,9 +108,9 @@ class LFRicArgDescriptor(Descriptor):
     # indicates either x1d or y1d.
     # Note, the LFRic infrastructure currently does not have 'region' as
```
Comment (arporter) [inline src/psyclone/domain/lfric/lfric_arg_descriptor.py:109]: Presumably this comment needs updating/removing?

File: src/psyclone/tests/dynamo0p3_test.py [hunk 3]
```diff
@@ -4302,20 +4301,54 @@ def test_single_stencil_literal(dist_mem, tmpdir):
     assert output6 in result
 
 
-def test_stencil_region_unsupported(dist_mem):
-    '''Check that we raise an exception if the value of the stencil type
-    in stencil(<type>[,<extent>]) is region. This is not a parse error
-    as region is a valid value, it is just that the LFRic
-    infrastructure does not yet support it. '''
+def test_stencil_region(dist_mem, tmpdir):
```
Comment (arporter) [inline src/psyclone/tests/dynamo0p3_test.py:4304]: Would you mind creating a new test file, `tests/domain/lfric/lfric_stencil_tests.py` and moving at least this test there. If you're feeling really generous you might move the other stencil tests too :-)

File: src/psyclone/tests/test_files/dynamo0p3/infrastructure/argument_mod.F90 [hunk 4]
```diff
@@ -161,7 +161,8 @@ module argument_mod
   integer, public, parameter :: X1D        = 2
   integer, public, parameter :: Y1D        = 3
   integer, public, parameter :: CROSS      = 4
-  integer, public, parameter :: CROSS2D    = 5
+  integer, public, parameter :: REGION     = 5
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/infrastructure/argument_mod.F90:164]: Please update the copyright date for the modifications to this file to "2020-2021"

File: src/psyclone/tests/test_files/dynamo0p3/testkern_stencil_region_mod.f90 [hunk 5]
```diff
@@ -31,10 +31,9 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Author R. W. Ford, STFC Daresbury Lab
-! Modified I. Kavcic, Met Office
+! Modified I. Kavcic, A. Coughtrie, Met Office
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/testkern_stencil_region_mod.f90:34]: Please update the copyright year range.

**Review by arporter — CHANGES_REQUESTED** [review]
Small, self-contained change.
All tests pass, including with compilation.
Examples all OK.
Updated UG builds fine.
Ref guide is fine.
All changes are pylint/pycodestyle clean.

Main request is for the addition of an example of a region stencil to the User Guide. Apart from that it's really just copyright dates and I've requested that you make a start on creating a separate test file for the stencil functionality (because dynamo0p3_test is a behemoth).
I don't think any of the submodules should be modified by this PR so please could you find a way of undoing those changes? Alternatively, perhaps you need to do a `git submodule update` and then commit the changes?

File: src/psyclone/tests/test_files/dynamo0p3/testkern_stencil_region_mod.f90 (hunk 5 above)
Comment (andrewcoughtrie) [reply src/psyclone/tests/test_files/dynamo0p3/testkern_stencil_region_mod.f90:34]: Done, see commit fdd34d286c25ea21986ea3520f7ae3d039b2937e

File: src/psyclone/tests/test_files/dynamo0p3/infrastructure/argument_mod.F90 (hunk 4 above)
Comment (andrewcoughtrie) [reply src/psyclone/tests/test_files/dynamo0p3/infrastructure/argument_mod.F90:164]: Done, see commit 77b51c2ebc81fb1860b1b0f15c9b48b73e20c70b

File: src/psyclone/tests/dynamo0p3_test.py (hunk 3 above)
Comment (andrewcoughtrie) [reply src/psyclone/tests/dynamo0p3_test.py:4304]: I've moved everything I believe to be a mainly stencil test to a new file, this doesn't include tests of things like halo exchange tests that are testing for stencil compatibility for instance. See commit 3306ebb8ff4b9499282e0f3adac6305b55b0612f

File: src/psyclone/domain/lfric/lfric_arg_descriptor.py (hunk 2 above)
Comment (andrewcoughtrie) [reply src/psyclone/domain/lfric/lfric_arg_descriptor.py:109]: Removed, see commit 6b2e780d510f5c50ed82875502eb3d0a74c7099a

File: doc/user_guide/dynamo0p3.rst (hunk 1 above)
Comment (andrewcoughtrie) [reply doc/user_guide/dynamo0p3.rst:1106]: Added an example of the REGION stencil to the user doc, see commit ae504ed6b45d1167d2d5c2c4b049a98e8f248954

**Comment by andrewcoughtrie** [comment]
Due to the issue of reversing the changes added to the externals directory i've created a new branch from the commit before the files were changed. This has resulted in the need for a new pull request (#1060), the changes in response to the review are included in the new PR, comments on the review in this PR include links to the commits for the changes. Hopefully this makes it clear what is going on and where the changes have moved to. 

If @arporter  is happy with how this has been handled this PR can be closed.

File: src/psyclone/tests/dynamo0p3_test.py (hunk 3 above)
Comment (arporter) [reply src/psyclone/tests/dynamo0p3_test.py:4304]: Wow, there was a lot! Thanks very much.

**Comment by arporter** [comment]
Closing this one as has been superseded by #1060.

**Comment by arporter** [comment]
@andrewcoughtrie is it OK if I delete the branch associated with this PR?

**Comment by andrewcoughtrie** [comment]
@arporter I'm not using the branch so it can be deleted yes.

**Pull request closed without merging**